### PR TITLE
wireless/bcm43xxx: add auto power saving support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

wireless/bcm43xxx: add auto power saving support

switch firmware power mode between PM_MAX/PM_FAST

## Impact

N/A

## Testing

bcm43013 iperf test

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 1]
```diff
@@ -778,6 +859,9 @@ static int bcmf_txavail(FAR struct net_driver_s *dev)
 {
   FAR struct bcmf_dev_s *priv = (FAR struct bcmf_dev_s *)dev->d_private;
 
+#ifdef CONFIG_IEEE80211_BROADCOM_LOWPOWER
+  bcmf_lowpower_poll(priv);
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:863]: put to work thread?

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 2]
```diff
@@ -742,6 +750,11 @@ static int bcmf_ifdown(FAR struct net_driver_s *dev)
 
   if (priv->bc_bifup)
     {
+      if (!work_available(&priv->lp_work))
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:753]: need guard by CONFIG_IEEE80211_BROADCOM_LOWPOWER

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 3]
```diff
@@ -755,6 +768,74 @@ static int bcmf_ifdown(FAR struct net_driver_s *dev)
...
+
+  if (!priv->bc_bifup)
+    {
+      return;
+    }
+
+  if (TICK2SEC(cticks - priv->lp_time) >=
+      CONFIG_IEEE80211_BROADCOM_LOWPOWER_TIMEOUT)
+    {
+      bcmf_wl_set_pm(priv, PM_MAX);
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:796]: where PM_MAX?

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 4]
```diff
@@ -755,6 +768,74 @@ static int bcmf_ifdown(FAR struct net_driver_s *dev)
...
+{
+  if (priv->bc_bifup)
+    {
+      priv->lp_time = clock_systime_ticks();
+      if (priv->lp_mode != PM_FAST)
+        {
+          bcmf_wl_set_pm(priv, PM_FAST);
+        }
+
+      if (work_available(&priv->lp_work) && !priv->lp_pending)
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:828]: lp_pending don't need, it's enough to check work_available

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c (hunk 1 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:865]: tx work thread will try again after receive data, it is more appropriate to poll low power in txavailable

https://github.com/apache/incubator-nuttx/blob/master/drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c#L226-L232

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c (hunk 2 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:754]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c (hunk 3 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:798]: https://github.com/apache/incubator-nuttx/blob/master/drivers/wireless/ieee80211/bcm43xxx/bcmf_ioctl.h#L1161-L1163

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c (hunk 4 above)
Comment (xiaoxiang781216) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:830]: If so, you can check lp_mode == PM_MAX

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 5]
```diff
@@ -755,6 +770,74 @@ static int bcmf_ifdown(FAR struct net_driver_s *dev)
...
+ * Input Parameters:
+ *   arg - context of device to use
+ *
+ ****************************************************************************/
+
+static void bcmf_lowpower_poll(FAR struct bcmf_dev_s *priv)
+{
+  if (priv->bc_bifup)
+    {
+      priv->lp_time = clock_systime_ticks();
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:824]: lp_time -> lp_tick

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 6]
```diff
@@ -755,6 +770,74 @@ static int bcmf_ifdown(FAR struct net_driver_s *dev)
...
+ * Input Parameters:
+ *   arg - context of device to use
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_IEEE80211_BROADCOM_LOWPOWER
+static void bcmf_lowpower_work(FAR void *arg)
+{
+  FAR struct bcmf_dev_s *priv = (FAR struct bcmf_dev_s *)arg;
+  clock_t cticks = clock_systime_ticks();
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:788]: cticks to ticks

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 7]
```diff
@@ -755,6 +770,74 @@ static int bcmf_ifdown(FAR struct net_driver_s *dev)
...
+
+  if (TICK2SEC(cticks - priv->lp_time) >=
+      CONFIG_IEEE80211_BROADCOM_LOWPOWER_TIMEOUT)
+    {
+      bcmf_wl_set_pm(priv, PM_MAX);
+      priv->lp_pending = false;
+    }
+  else
+    {
+      cticks = SEC2TICK(CONFIG_IEEE80211_BROADCOM_LOWPOWER_TIMEOUT) -
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:803]: let's calculate the delta before if, so we can reuse the result at here and line 795

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 8]
```diff
@@ -755,6 +770,74 @@ static int bcmf_ifdown(FAR struct net_driver_s *dev)
...
+ *   arg - context of device to use
+ *
+ ****************************************************************************/
+
+static void bcmf_lowpower_poll(FAR struct bcmf_dev_s *priv)
+{
+  if (priv->bc_bifup)
+    {
+      priv->lp_time = clock_systime_ticks();
+      if (priv->lp_mode != PM_FAST)
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:825]: move the check to bcmf_wl_set_pm?

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c (hunk 6 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c]: Done

**Review by pkarashchenko — APPROVED** [review]

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 9]
```diff
@@ -426,6 +426,34 @@ int bcmf_driver_download_clm(FAR struct bcmf_dev_s *priv)
...
+  /* Set default power save mode */
+
+#ifdef CONFIG_IEEE80211_BROADCOM_LOWPOWER
+  if (priv->lp_mode != mode)
+#endif
+    {
+      out_len = 4;
+      value   = mode;
+      ret = bcmf_cdc_ioctl(priv, interface, true, WLC_SET_PM,
+                           (uint8_t *)&value, &out_len);
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:445]: Thanks for review, let me change the mode type to uint8_t

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c]: Yes, I think it's better to create a new PR to track the endian issue.

**Pull request merged**